config.go

```diff
@@ -13,45 +13,81 @@ import (
 	"strings"
 	"time"
 
-	"github.com/BurntSushi/toml"
 	log "github.com/sirupsen/logrus"
+	"github.com/troian/toml"
 )
 
+// Config instantiation must be done in multiple steps
```

Comment:
Why is this here?

Reply:
A bunch of commented out code just causes confusion.
Add a small comment in the code where the initialisation is done to describe why it needs to be done in this order. Wrap all these steps in a function `HandleConfig` or similar.
I think there are better ways than having this big block of commented out code.

Reply:
Well, this is common approach for Go to provide an example of API usage.
That will look quite ugly

```go
ca := cagent.New()
ca.HandleConfig()
```

begs to

```go
func LoadConfig(file string) *Cagent {
	ca := New()

	err := ca.ReadConfigFromFile(file)
	ca.ApplyEnv()
	if os.IsNotExist(err) {
		// this is ok
		err = ca.CreateDefaultConfigFile(file)
		if err != nil {
			log.Fatal(err)
		}
	} else if err != nil {
		if strings.Contains(err.Error(), "cannot load TOML value of type int64 into a Go float") {
			log.Fatalf("Config load error: please use numbers with a decimal point for numerical values")
		} else {
			log.Fatalf("Config load error: %s", err.Error())
		}
	}

	if err = ca.Initialize(); err != nil {
		log.Fatal(err)
	}

	return ca
}
```

Reply:
> Well, this is common approach for Go to provide an example of API usage.

Agree, but we don't document an API here for external developers. This is bootstrap code for the application and won't be used in any other place.

Another note to your example: The function is called `LoadConfig ` but also does instantiate `ca`. Functions that instantiate things internally are always hard to test. Create `ca` outside the function and pass it as an argument. Check google for `Dependency Injection`.

I image the whole config flow as following:
1. Try to read config file and set values in `ca`
2. If no config is present, try to create default config at location and load it. If this fails panic
3. Parse environment vars and overwrite values in `ca`
4. Check which flags are provided and overwrite values in `ca` in case

Each step can basically live in its own function and all steps can be combined in a helper function like `HandleConfig` or something similar named to keep `main` more clean.

Reply:
ok